core/src/main/java/com/linecorp/armeria/common/stream/ByteBufDeframerInput.java

```diff
@@ -133,28 +140,43 @@ public ByteBuf readBytes(int length) {
 
     private ByteBuf readBytesSlow(int length) {
         ByteBuf value = null;
-        int remained = length;
-        for (ByteBuf buf : queue) {
+        int remaining = length;
+        for (final Iterator<ByteBuf> it = queue.iterator(); it.hasNext();) {
+            final ByteBuf buf = it.next();
             if (!buf.isReadable()) {
+                it.remove();
+                buf.release();
                 continue;
             }
 
-            final int readSize = Math.min(remained, buf.readableBytes());
-            if (readSize == length) {
-                return buf.readRetainedSlice(readSize);
-            } else {
-                if (value == null) {
-                    value = alloc.buffer(length);
+            final int readableBytes = buf.readableBytes();
```

Comment:
You could move this line before the `if (!buf.readable())` above, and change the `if` statement to `if (readableBytes == 0)`

Reply:
I cleaned up ByteBufDeframerInput a bit to immediately remove a `ByteBuf` from `queue`  when it can no longer be read.